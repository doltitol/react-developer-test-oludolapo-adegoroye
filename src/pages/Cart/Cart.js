import React, { PureComponent } from 'react';
import './cart.style.scss';
import { connect } from 'react-redux';
import { addQuantity, removeQuantity } from '../../redux/actions/cartActions';
import { numberCommaFormatter } from '../../util';
import MainButton from '../../components/Buttons/MainButton';
import CartItem from '../../components/CartItem/CartItem';


export class Cart extends PureComponent {
  getTotal = () => {
    const tax = 21;
    const total = this.props.cartItems.reduce(
      (acc, cur) =>
        acc +
        cur.prices.filter(
          (price) => price.currency.symbol === this.props.currency
        )[0].amount *
          cur.qty,
      0
    );
    const taxAmount = total * (tax / 100);
    return {
      tax,
      taxAmount,
      total,
    };
  };
  render() {
    return (
      <div className='cart' ref={this.cartRef}>
        <h1 className='cart-heading'>cart</h1>
        <div className='cart-cartList'>
          {this.props.cartItems.length === 0 && (
            <div className='cart-cartList-empty'>
              <h1 data-testid='empty-text'>Your cart is empty!</h1>
            </div>
          )}
          {this.props.cartItems.map((item, index) => (
            <CartItem cartItem={item} key={`${item.id}-${index}`} />
          ))}
        </div>
        <div className='cart-cartTotal'>
          <div className='cart-cartTotal-item'>
            <div className='cart-cartTotal-item-heading'>
              <p>Tax {this.getTotal().tax}%:</p>
            </div>
            <div className='cart-cartTotal-item-amount'>
              <p>
                {this.props.currency}
                {numberCommaFormatter(this.getTotal().taxAmount.toFixed(2))}
              </p>
            </div>
          </div>
          <div className='cart-cartTotal-item'>
            <div className='cart-cartTotal-item-heading'>
              <p>Quantity</p>
            </div>
            <div className='cart-cartTotal-item-amount'>
              <p>{this.props.cartQuantity}</p>
            </div>
          </div>
          <div className='cart-cartTotal-item'>
            <div className='cart-cartTotal-item-heading'>
              <p>Total</p>
            </div>
            <div className='cart-cartTotal-item-amount'>
              <p>
                {this.props.currency}
                {numberCommaFormatter(this.getTotal().total.toFixed(2))}
              </p>
            </div>
          </div>
          <div className='cart-cartTotal-button'>
            <MainButton
              text='order'
              onClick={() => {}}
              height='43px'
              width='279px'
            />
          </div>
        </div>
      </div>
    );
  }
}
const mapStateToProps = (state) => {
  return {
    cartItems: state.cart.cartItems,
    currency: state.cart.currency,
    cartQuantity: state.cart.cartQuantity,
    total: state.cart.total,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    addQuantity: (product) => {
      dispatch(addQuantity(product));
    },
    removeQuantity: (product) => {
      dispatch(removeQuantity(product));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
