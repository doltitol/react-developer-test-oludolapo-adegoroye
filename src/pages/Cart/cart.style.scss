@import '../../styles/index.scss';

.cart {
  @include container;
  &-heading {
    text-transform: uppercase;
    @include raleway(32px, 40px, 700);
    margin-top: 50px;
  }
  &-cartList {
    margin-top: 50px;
    border-top: 1px solid $border;
  }
  &-cartTotal {
    margin: 20px 0;
    &-item {
      margin: 5px 0;
      display: flex;
      flex-direction: row;
      &-heading {
        @include raleway(20px, 28px, 400);
        width: 100px;
      }
      &-amount {
        @include raleway(20px, 28px, 700);
      }
    }
    &-button{
    margin-top: 20px;
    }
  }
}
