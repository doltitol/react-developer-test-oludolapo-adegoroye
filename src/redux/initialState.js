export const initialState = {
  cartItems: JSON.parse(localStorage.getItem('cartItems') || '[]'),
  cartQuantity: JSON.parse(localStorage.getItem('cartItems') || '[]').reduce(
    (acc, item) => acc + item.qty,
    0
  ),
  currency: '$',
};
