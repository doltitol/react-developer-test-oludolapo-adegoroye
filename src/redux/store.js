import { createStore, combineReducers } from 'redux';
import cartReducer from './reducers/cartReducers';

const rootReducer = combineReducers({
  cart: cartReducer,
});
export const store = createStore(rootReducer);
