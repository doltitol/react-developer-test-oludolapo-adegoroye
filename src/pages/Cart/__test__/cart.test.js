import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import { store } from '../../../redux/store';
import Cart from '../Cart';

describe('<Cart /> spec', () => {
  it('renders the empty component', () => {
    const cartItems = [];
    const container = render(
      <Provider store={store}>
        <Cart cartItems={cartItems} />
      </Provider>
    );
    const getByTestId = container.getByTestId;
    expect(getByTestId('empty-text').textContent).toEqual(
      'Your cart is empty!'
    );
    expect(container.children).toMatchSnapshot();
  });
});
