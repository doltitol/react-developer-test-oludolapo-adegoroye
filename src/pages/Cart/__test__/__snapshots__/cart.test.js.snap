// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> spec renders the empty component 1`] = `undefined`;

exports[`<Cart /> spec renders the empty component 2`] = `undefined`;
